Add note for changing IPAM settings (#27090)
cilium-team/cilium: Documentation/network/concepts/ipam/index.rst

Dynamic CIDR/IP allocation   ❌                     ❌                       ✅          ✅          ✅              ✅                ❌
============================ ====================== ======================== =========== =========== =============== ================= ===============

> Changing the IPAM mode of an existing cluster is impossible.

Reviewer: Maybe..?

```suggestion
Changing the IPAM mode of an existing cluster is strongly discouraged as it may cause persistent connectivity disruption for existing workloads.
```

Reviewer: This plus the "you must reinstall Cilium ..." should make it clear both why and also how to make the change if necessary.

Reviewer: (I prefer my proposal [below](https://github.com/cilium/cilium/pull/27090#discussion_r1276688974) instead, but I'm also open to this proposal. In particular I think that the previous text that states "it is impossible" followed by "here's how to do it" is confusing.)

You must reinstall Cilium to change the IPAM mode.

> Before reinstalling Cilium, delete all CiliumNode CRD objects.

Reviewer: I struggle a little bit with the step to delete all CiliumNode CRD objects. Exactly when should they delete the CilumNode CRD objects? Before uninstalling? After uninstalling?

I think that the intent here is to state what is necessary to switch modes, but at the same time it also feels like it's half way between documenting limitations and trying to provide a migration guide. 

Strictly speaking this is probably correct, but also if we want to provide a path for users to change the IPAM mode then we could remove the need to delete CiliumNodeCRD objects in the implementation, and probably we would come up with a more comprehensive migration guide rather than dropping three terse sentences on how it could be done.

Author: Yeah, I agree.

> Changing the IPAM mode of an existing cluster is impossible.
> You must reinstall Cilium to change the IPAM mode.
> Before reinstalling Cilium, delete all CiliumNode CRD objects.

Reviewer: Here's another alternative proposal:
```suggestion
Do not change the IPAM mode of an existing cluster. Changing the IPAM mode in
a live environment may cause persistent connectivity disruption for existing workloads.
The safest path to change IPAM mode is to install a fresh Kubernetes cluster with the new IPAM configuration.
```

I feel this would make it clear that the Cilium community has not attempted to solve this problem yet, and if we want to solve this problem then we can open issues/CFPs/PRs and work through the process to provide a path.

Author: Is a fresh cluster needed? I guess removing Cilium and run https://docs.cilium.io/en/stable/cmdref/cilium_cleanup/ will do the trick?

Reviewer: Not quite, mainly because that cilium cleanup only cleans local node state, not state in k8s, so the step that was referenced earlier about deleting CiliumNode resources won't happen AFAIK.

Author: Yes, that's also my understanding. I'm refering to "install a fresh cluster" So in theory deleting CiliumNode CRD + run the cleanup.

Reviewer: Ah yes that should technically work as well. We could alternatively describe it as completely removing Cilium from the cluster including all custom k8s resources and then freshly installing Cilium.

Author: Yes, I think that's the best approach. Then we can link from "changing IPAM mode" to this section. I could work on that next week. Maybe you can give me some hints on where I should add that.

Reviewer: Ah, what I had in mind was just to tersely describe here that Cilium must be completely removed and reinstalled. But yeah it seems reasonable to document that too. I'm not seeing an obvious place in the hierarchy for this right now, maybe @zacharysarah might have a thought on this?

Reviewer 2: Would it make sense to add that heading instead under troubleshooting? It's not explicitly a troubleshoot per se, so that's something to consider. We could also add a heading for `Known issues`.

Minor tweaks to @joestringer's proposed update:
```suggestion
Don't change the IPAM mode of an existing cluster. Changing the IPAM mode in
a live environment may cause persistent disruption of connectivity for existing workloads.
The safest path to change IPAM mode is to install a fresh Kubernetes cluster with the new IPAM configuration.
```

> I feel this would make it clear that the Cilium community has not attempted to solve this problem yet, and if we want to solve this problem then we can open issues/CFPs/PRs and work through the process to provide a path.

Would adding explicit language to that effect be useful, or is the statement sufficiently self-disclosing in context? (I think your language here would work great, with some tweaks.)

Author: implemented the recommendations

Reviewer: > > I feel this would make it clear that the Cilium community has not attempted to solve this problem yet, and if we want to solve this problem then we can open issues/CFPs/PRs and work through the process to provide a path.
>
> Would adding explicit language to that effect be useful, or is the statement sufficiently self-disclosing in context? (I think your language here would work great, with some tweaks.)

I think that it would help to call this out explicitly. Even though it's Open Source and hence there's always an opportunity to change the way things work, I think that it's often difficult for users to understand whether something is impossible by design or just that no-one solved it yet. I'll propose some wording to this effect.

> 

Reviewer: ```suggestion
If you are interested in extending Cilium to support migration between IPAM modes, see :gh-issue:`27164`.
```

   gke
   crd
   deep_dive

.. note::

> 

Reviewer: How does this render?

Again I think this is a case where `.. note::` is unsuitable. Either create a section title for modifying IPAM configuration or just embed the text directly without a note.

>     Changing the IPAM mode of an existing cluster is not supported. 
>     Cilium must be reinstalled to change the IPAM mode. Before reinstalling
>     Cilium, make sure that all CiliumNode CRD objects are deleted.

Reviewer 2: s/supported/recommended per ☝🏻, use active voice
```suggestion
    Changing the IPAM mode of an existing cluster isn't recommended. 
    You must reinstall Cilium to change the IPAM mode. 
    
    Before reinstalling Cilium, delete all CiliumNode CRD objects.
```
